Broadcast shouldn't use inbounds when calling f

mbauman opened the pull request:
Previously, `broadcast(getindex, A, I...)` would call `getindex` within an
at-inbounds context. This patch simply moves the function call so that it
will throw a BoundsError appropriately.

Fixes #19203.

mbauman committed:
Broadcast shouldn't use inbounds when calling f

Previously, `broadcast(getindex, A, I...)` would call `getindex` within an
at-inbounds context. This patch simply moves the function call so that it
will throw a BoundsError appropriately.

mbauman commented on test/broadcast.jl:
```diff
@@ -337,3 +337,8 @@ end
 @test broadcast(+, 1.0, (0, -2.0), [1]) == [2.0, 0.0]
 @test broadcast(*, ["Hello"], ", ", ["World"], "!") == ["Hello, World!"]
+
+# Ensure that broadcast doesn't use @inbounds when calling the function
```
The tests are run with `—check-bounds=yes`, no?  What's the best way to make sure that this catches regressions?

yuyichao replied:
`test/boundscheck.jl`

stevengj commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19188#issuecomment-258159314) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/6ff48f5a63af0e8a19f29066fc3140be0c1001f1/7cade9f_vs_421f079/report.md). cc @jrevels

stevengj commented:
There seem to be some relevant slowdowns.  It seems like we should still be able to use `@inbounds` somewhere, maybe in `_broadcast_getindex`?

mbauman commented:
This is a bug-fix.  The loss in performance is entirely due to restoring bounds checks in places where they had been erroneously removed before.  In order to re-gain that performance, we'd have to add `@propagate_inbounds` annotations to the entire call chain, and then annotate the broadcast calls — if they're safe —  with `@inbounds f.(...)`.

Unfortunately, this uses `@noinline` (I assume) to ensure type stability of the inner loop, which precludes the use of `@propagate_inbounds`.

stevengj commented:
But wouldn't adding
```jl
@inline _broadcast_getindex(::Any, A::AbstractArray, I) = @inbounds A[I]
```
be safe?  Here, the indices `I` are constructed by `broadcast` itself, and should always be inside the array bounds.

mbauman commented:
Sure, but that's unrelated to this change.

mbauman committed:
Add inbounds elsewhere in broadcast

Try to restore some performance by allowing inbounds propagation into _broadcast_getindex.

mbauman commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19188#issuecomment-258258062) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/aa1e61f1dc506fb67afbc07a06187b409d9e8cc3/b4bf141_vs_421f079/report.md). cc @jrevels

tkelman commented:
I don't think this is relevant on release-0.5 without https://github.com/JuliaLang/julia/pull/16986
